### PR TITLE
Support for SerializerMethodField

### PR DESCRIPTION
**Pull request by therefromhere**
First attempt at support for SerializerMethodField (#137).

Added support for hinting the schema for a SerializerMethodField in 2 ways:

* Added a decorator for the use case where the serializer method is using a serializer: `swagger_serializer_method(serializer_class)`
* Added support for python3.5+ return type hinting, for the use case where the serializer method is returning a type that we have mapped to a swagger type.

TODO:

- [x] Implement inspect based on return type hint
- [x] Fix tests on versions that don't support `->` return type hinting syntax (<py35)
- [x] Don't throw away the SerializerMethod() params (help_text etc)
- [x] Mark the decorator result as readOnly
- [x] Support more types for return type hint
- [x] Add docs/changelog

Future improvements (probably out of scope for this PR):

* More support for type hinting? (eg `-> List[UUID]`)
* Support for providing extra return type hints in settings

### CODE REVIEW

File: src/drf_yasg/inspectors/field.py [hunk 1]
```diff
@@ -409,6 +411,86 @@ def get_basic_type_info(field):
     return result
 
 
+def decimal_return_type():
+    return openapi.TYPE_NUMBER if rest_framework_settings.COERCE_DECIMAL_TO_STRING else openapi.TYPE_STRING
```
[the remark this answers is not in the document]
Comment (therefromhere) [reply src/drf_yasg/inspectors/field.py]: yep, fixed

File: src/drf_yasg/inspectors/field.py [hunk 2]
```diff
@@ -409,6 +411,86 @@ def get_basic_type_info(field):
...
+            if sys.version_info >= (3, 5):
+                # look for Python 3.5+ style type hinting of the return value
+                return_class = inspect.signature(method).return_annotation
+
+                if not issubclass(return_class, inspect._empty):
+                    type_info = get_basic_type_info_from_type(return_class)
+
+                    SwaggerType, ChildSwaggerType = self._get_partial_types(field, swagger_object_type, use_references,
+                                                                            **kwargs)
+                    return SwaggerType(**type_info)
```
[the remark this answers is not in the document]
Comment (therefromhere) [reply src/drf_yasg/inspectors/field.py]: the comments seem to have been moved due to later commits. I think this is done

File: src/drf_yasg/utils.py [hunk 3]
```diff
@@ -170,6 +170,21 @@ def decorator(view_method):
...
+def swagger_serializer_method(serializer_class):
+    """
+    Decorate the method for a serializers.SerializerMethodField
+    :param serializer_class:
+    :return:
+    """
+
+    def decorator(serializer_method):
+        # stash the serializer class for SerializerMethodFieldInspector to find
+        serializer_method.swagger_serializer_class = serializer_class
```
[the remark this answers is not in the document]
Comment (therefromhere) [reply src/drf_yasg/utils.py]: Agreed, I've updated to allow class or instance

File: testproj/users/serializers_old_py.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    hinted_number = serializers.SerializerMethodField(
+        help_text="the type hint on the method should determine this to be a number")
+
+    non_hinted_number = serializers.SerializerMethodField(
+        help_text="No hint on the method, so this is expected to fallback to string")
+
+    # `->` syntax isn't supported, instead decorate with a serializer that returns the same type
+    # (a bit of a hack, but means we get the same output)
+    @swagger_serializer_method(serializer_class=serializers.FloatField)
```
[the remark this answers is not in the document]
Comment (therefromhere) [reply testproj/users/serializers_old_py.py]: fixed in later commit

File: src/drf_yasg/inspectors/field.py [hunk 5]
```diff
@@ -409,6 +411,86 @@ def get_basic_type_info(field):
...
+
+        if isinstance(field, serializers.SerializerMethodField):
+            method = getattr(field.parent, field.method_name)
+
+            if hasattr(method, "swagger_serializer_class"):
+                # attribute added by the swagger_serializer_method decorator
+
+                serializer_kwargs = {
+                    # a bit hacky? copy help_text from the SerializerMethodField
+                    "help_text": field.help_text,
```
[the remark this answers is not in the document]
Comment (therefromhere) [reply src/drf_yasg/inspectors/field.py]: Seems reasonable - so there's maybe 3 places help_text could come from:

a) ```SerializerMethodField(help_text)```
b) ```@swagger_serializer_method(serializer=OtherStuffSerializer(help_text))``` (if an instance of serializer was passed)
c) the get_blah method's docstring.

Is that a reasonable priority order?  Or should it be b, then, a, then finally fallback to c?

[the remark this answers is not in the document]
Comment (axnsan12) [reply src/drf_yasg/inspectors/field.py]: Your order makes most sense to me

[the remark this answers is not in the document]
Comment (therefromhere) [reply src/drf_yasg/inspectors/field.py]: done.

**Comment by therefromhere** [comment]
I'll create a fresh squashed branch to tidy the history.

**Comment by therefromhere** [comment]
Closing in favour of squashed version, #179

**Comment by axnsan12** [comment]
There was no need for that, I always squash and merge anyway :smiley:

**Pull request closed without merging**